### PR TITLE
Annotate System.Security.Cryptography.Pkcs for nullable

### PR DESCRIPTION
**Pull request by buyaa-n**
Contributes to https://github.com/dotnet/runtime/issues/2339

### CODE REVIEW

File: src/libraries/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/AnyOS/ManagedPal.KeyAgree.cs
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/AnyOS/ManagedPal.KeyAgree.cs]: What's the TODO? Whose action is it?

File: src/libraries/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/PkcsPalWindows.cs
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/PkcsPalWindows.cs]: I don't think the comment has long-term value.

Doesn't `[MaybeNull]` solve this problem?  (I'm not sure why the generic constraint can't be applied; but if it can't, [MaybeNull] seems like it solves the return type nullability).

File: src/libraries/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/AnyOS/ManagedPal.KeyAgree.cs
Comment (buyaa-n) [reply src/libraries/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/AnyOS/ManagedPal.KeyAgree.cs]: Mine 😄, as you can see `if (rid.RKeyId.Value.Other.Value.KeyAttr != null)` is checked above but compiler still complaining, wondering if this is a but for nullable value type and if it is known issue CC @stephentoub @safern  @eiriktsarpalis

File: src/libraries/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/PkcsPalWindows.cs
Comment (buyaa-n) [reply src/libraries/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/PkcsPalWindows.cs]: That comment is my question, not intended to keep, yes i can use [MaybeNull] just thought its better to make it restricted to a class and make nullable

**Review by bartonjs — APPROVED** [review]

**Comment by jkoritzinsky** [comment]
I'm seeing local build failures as well as official build failures due to this PR. (example official build failure: https://dev.azure.com/dnceng/internal/_build/results?buildId=529871&view=results).

**Comment by jkoritzinsky** [comment]
cc @dotnet/runtime-infrastructure

**Comment by buyaa-n** [comment]
> I'm seeing local build failures as well as official build failures due to this PR. (example official build failure: https://dev.azure.com/dnceng/internal/_build/results?buildId=529871&view=results).

Sorry, working on it

**Comment by jaredpar** [comment]
Seeing the same. The errors are really straight forward too:

```
Internal/Cryptography/Pal/AnyOS/ManagedPal.cs(35,39): error CS8600: (NETCORE_ENGINEERING_TELEMETRY=Build) Converting null literal or possible null value to non-nullable type.
Internal/Cryptography/Pal/AnyOS/ManagedPal.cs(75,27): error CS8600: (NETCORE_ENGINEERING_TELEMETRY=Build) Converting null literal or possible null value to non-nullable type.
Internal/Cryptography/Pal/AnyOS/ManagedPal.cs(75,24): error CS8600: (NETCORE_ENGINEERING_TELEMETRY=Build) Converting null literal or possible null value to non-nullable type.
Internal/Cryptography/Pal/AnyOS/ManagedPal.cs(75,24): error CS8603: (NETCORE_ENGINEERING_TELEMETRY=Build) Possible null reference return.
Internal/Cryptography/Pal/AnyOS/ManagedPal.cs(77,27): error CS8600: (NETCORE_ENGINEERING_TELEMETRY=Build) Converting null literal or possible null value to non-nullable type.
Internal/Cryptography/Pal/AnyOS/ManagedPal.cs(77,24): error CS8600: (NETCORE_ENGINEERING_TELEMETRY=Build) Converting null literal or possible null value to non-nullable type.
Internal/Cryptography/Pal/AnyOS/ManagedPal.cs(77,24): error CS8603: (NETCORE_ENGINEERING_TELEMETRY=Build) Possible null reference return.
Internal/Cryptography/Pal/AnyOS/ManagedPal.cs(80,27): error CS8600: (NETCORE_ENGINEERING_TELEMETRY=Build) Converting null literal or possible null value to non-nullable type.
Internal/Cryptography/Pal/AnyOS/ManagedPal.cs(80,24): error CS8600: (NETCORE_ENGINEERING_TELEMETRY=Build) Converting null literal or possible null value to non-nullable type.
```

This seems like a straight build issue. How is it that our official and PR builds can differ in the results here? That seems like a fairly critical failure. Two likely causes:

1. Official and PR builds are using different compiler toolsets
1. This is code that doesn't build in PR

**Comment by jaredpar** [comment]
Note: the PR was 100% green here so this is either a test gap or infra disconnect on toolsets.

**Comment by buyaa-n** [comment]
I see dependent PR merged just before my merge and we have lowered nullability error to warning on CI builds

**Comment by jaredpar** [comment]
> I see dependent PR merged just before my merge and we have lowered nullability error to warning on CI builds

The official builds have nullability as error. Having nullability be error in official and warning everywhere else is just going to keep leading to problems like this. Need to be consistent here.

**Comment by jaredpar** [comment]
Verified the toolset is consistent between the internal and PR builds.

**Comment by jaredpar** [comment]
@buyaa-n where did we make the change to have nullable diagnostics be warnings?

**Comment by stephentoub** [comment]
> Having nullability be error in official and warning everywhere else is just going to keep leading to problems like this. Need to be consistent here.

It's needs to be error, everywhere.

**Comment by buyaa-n** [comment]
> @buyaa-n where did we make the change to have nullable diagnostics be warnings?

Seems  we disabled warnAsError on Windows in CI, the PR is here https://github.com/dotnet/runtime/pull/32515

**Comment by jaredpar** [comment]
This is hitting PRs now too

https://dev.azure.com/dnceng/public/_build/results?buildId=529965

**Comment by stephentoub** [comment]
The root problem here is a PR to null-annotate library A that depended on library B, which wasn't null-annotated yet... the PR for annotating B got merged just before the one for A, so it's the typical two-PRs-with-conflicts-crossing-in-CI issue.

**Comment by jaredpar** [comment]
Yep. That seems to be the case.

**Comment by stephentoub** [comment]
Switched the nullable mode for the project to unblock CI:
https://github.com/dotnet/runtime/pull/32617

**Comment by buyaa-n** [comment]
> The root problem here is a PR to null-annotate library A that depended on library B, which wasn't null-annotated yet... the PR for annotating B got merged just before the one for A, so it's the typical two-PRs-with-conflicts-crossing-in-CI issue.

Yes that is the case, I thought even it was merged in time I wouldn't see the error, but seems its warn only on Windows CI, so I might have seen error in other CI legs

**Comment by buyaa-n** [comment]
> Switched the nullable mode for the project to unblock CI: #32617

@stephentoub  raised PR for build fix, please revew

**Comment by safern** [comment]
> It needs to be error, everywhere.

This is failing in OSX in the official build. WarnAsErrors are disabled on windows in both CI and official because of: https://github.com/dotnet/runtime/issues/31888

I'm working with @MattGal on using the right pools with the necessary version of CMake we need in order to be able to enable WarnAsErrors again on Windows. Since I'm going OOF tomorrow, it might be good if someone else could follow up on that, maybe @ViktorHofer or @jkoritzinsky ?

**Pull request merged**
